### ISSUE TITLE
Provide certificate-based authentication through attributes

### ISSUE THREAD

**Issue by tomkerkhove** [arcus-azure/arcus.webapi#31]
**Is your feature request related to a problem? Please describe.**
Quite often we need to enforce tight security on an API where certificates are often used.

**Describe the solution you'd like**
A declarative way to tag operations on API controllers with an Authorize attribute that defines what client certificates that are allowed.

**Comment by MassimoC** [comment]
An example of certificate validation can be found here: https://docs.microsoft.com/en-us/azure/app-service/app-service-web-configure-tls-mutual-auth

**Comment by stijnmoreels** [comment]
Could be fun to play with. As test we should generate those certificates with different subject names, expiration dates, ...

@tomkerkhove is this somewhat the same approach as the **Shared Access Key** attributes?

**Comment by tomkerkhove** [comment]
Yes, but it all depends on the way it should work and required configuration so I think we need more insights first.

**Comment by fgheysels** [comment]
> An example of certificate validation can be found here: https://docs.microsoft.com/en-us/azure/app-service/app-service-web-configure-tls-mutual-auth

A bit disappointing that the ASP.NET sample is using some old ASP.NET WebForms code.  That's like, 15 years old or so. :)

**Comment by fgheysels** [comment]
> **Is your feature request related to a problem? Please describe.**
> Quite often we need to enforce tight security on an API where certificates are often used.
> 
> **Describe the solution you'd like**
> A declarative way to tag operations on API controllers with an Authorize attribute that defines what client certificates that are allowed.

So, how do you see this in action / used ?  Something like this :

```csharp
[ClientCertificateValidation(CertificateThumbPrint="abcdefghi")]
public class MyController
{
}
```
?
I am doing something similar in another project (no client cert validation, but some additional/custom authorization), so I might be fiddling around with it a bit more to get something like the above working.

**Comment by tomkerkhove** [comment]
There are a couple of ways to achieve this and I'm open for feedback!

Personally, I would use the same approach like the shared access key attribute where you can either use an attribute or plug it in the middleware.

**Comment by fgheysels** [comment]
Indeed, I was also wondering if you want to have it as fine grained as this.  I think it would be useful if you could just plug it in as middleware so that the certificate validation is done on the entire api instead of having to apply it on each controller separately.

**Comment by tomkerkhove** [comment]
After discussion we see value in the following approaches:
- **Middleware** to enforce this on all operations which should be used as a default approach
- **Attributes** in case you only want to secure a subset of endpoints with this approach
  - Example - We want to receive webhooks and don't have an API gateway in front.
    - We have to lower our security on the endpoints which will accept webhooks
    - All the other endpoints use attributes.

Given the nature of those, they should share the validation core but should otherwise be separated given they have a different decoration approach.

Given they are separate the middleware is tracked in #39.

**Comment by ToonVanhoutte** [comment]
There are quite some things that can be validated: root CA, thumbprint, issuer, subject... You can validate the certificate itself, or the complete chain.

> Within Azure services itself, it would be nice if you can refer to a certificate in Key Vault.  This allows rotating etc...
> For other scenarios, you could provide the things you want to valdiate: thumbprint, subject, etc...

**Comment by stijnmoreels** [comment]
> After discussion we see value in the following approaches:
> 
> * **Middleware** to enforce this on all operations which should be used as a default approach
> * **Attributes** in case you only want to secure a subset of endpoints with this approach
>   
>   * Example - We want to receive webhooks and don't have an API gateway in front.
>     
>     * We have to lower our security on the endpoints which will accept webhooks
>     * All the other endpoints use attributes.
> 
> Given the nature of those, they should share the validation core but should otherwise be separated given they have a different decoration approach.
> 
> Given they are separate the middleware is tracked in #39.

@tomkerkhove by "middleware" do you mean a filter? Since filters in ASP.NET Core can be both used as attributes and globally registered.

https://docs.microsoft.com/en-us/aspnet/core/mvc/controllers/filters?view=aspnetcore-2.2

**Comment by tomkerkhove** [comment]
I was more talking about the following :
```csharp
using Microsoft.AspNetCore.Builder;

namespace Culture
{
    public static class RequestCultureMiddlewareExtensions
    {
        public static IApplicationBuilder UseRequestCulture(
            this IApplicationBuilder builder)
        {
            return builder.UseMiddleware<RequestCultureMiddleware>();
        }
    }
}
```
_https://docs.microsoft.com/en-us/aspnet/core/fundamentals/middleware/write?view=aspnetcore-2.2#middleware-extension-method_

But it's just an approach the essence was that it can be enforced per operation, endpoint or for all endpoints. How it is achieved is technical detail and open for discussion.

**Comment by tomkerkhove** [comment]
> There are quite some things that can be validated: root CA, thumbprint, issuer, subject... You can validate the certificate itself, or the complete chain.
> 
> * Within Azure services itself, it would be nice if you can refer to a certificate in Key Vault.  This allows rotating etc...
> * For other scenarios, you could provide the things you want to valdiate: thumbprint, subject, etc...

The requirement to have the certificate itself I think depends on what you want to validate. For example, if it's just checking the thumbprint, I don't need access to the cert.

However, if we need the cert it should be retrieved via `ISecretProvider` so that the consumer can do whatever they want.

I think we need to be pragmatic and start validating the following attributes:
- Thumbprint
- Subject
- Expiration date

Does everybody agree?

**Comment by stijnmoreels** [comment]
> > There are quite some things that can be validated: root CA, thumbprint, issuer, subject... You can validate the certificate itself, or the complete chain.
> > 
> > * Within Azure services itself, it would be nice if you can refer to a certificate in Key Vault.  This allows rotating etc...
> > * For other scenarios, you could provide the things you want to valdiate: thumbprint, subject, etc...
> 
> The requirement to have the certificate itself I think depends on what you want to validate. For example, if it's just checking the thumbprint, I don't need access to the cert.
> 
> However, if we need the cert it should be retrieved via `ISecretProvider` so that the consumer can do whatever they want.
> 
> I think we need to be pragmatic and start validating the following attributes:
> 
> * Thumbprint
> * Subject
> * Expiration date
> 
> Does everybody agree?

You can assign me to this if you like.

**Comment by tomkerkhove** [comment]
Thanks I will, but will wait so that people can give their input on my proposal.

/cc @ToonVanhoutte  @SamVanhoutte @fgheysels @MassimoC

**Comment by ToonVanhoutte** [comment]
Agree, but would add also Issuer to the list, as it is a common one!

**Comment by tomkerkhove** [comment]
Updated!

**Comment by MassimoC** [comment]
> I think we need to be pragmatic and start validating the following attributes:
> 
> * Thumbprint
> * Subject
> * Expiration date
> * Issuer
> 
> Does everybody agree?

Yes. BTW, I would love to have a fluent configuration of the middleware.
Probably the middleware will be the most used approach - in AppService, when u set'clientCertEnabled' to true, is for all the operations.

**Comment by tomkerkhove** [comment]
We tend to provide fluent builders for most of them indeed!

**Comment by tomkerkhove** [comment]
Closed via #43
